### PR TITLE
apf PostStartHook: handle error from PollImmediateUntil

### PR DESCRIPTION
**Pull request by tkashem**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
We should not attempt creation of mandatory objects if ensuring the suggested configuration resulted in an error.
We rely on the presence of the `exempt` priority level configuration object in the cluster to indicate whether we should ensure
suggested configuration.

It is unlikely to happen, but given the right condition the mandatory configuration can be created:
- `wait.PollImmediateUntil` for the suggested configuration does one try of the suggested configuration and it fails
- `stopCh` is closed
- `wait.PollImmediateUntil` for the suggested configuration returns an `ErrWaitTimeout`
- `wait.PollImmediateUntil` for the mandatory configuration, first attempt succeeds.

I think we should always handle error from `wait.PollImmediateUntil`

#### Does this PR introduce a user-facing change?
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go
Comment (MikeSpreitzer) [inline pkg/registry/flowcontrol/rest/storage_flowcontrol.go:134]: > // This only happens when the stop channel is closed.

**Review by MikeSpreitzer — APPROVED** [review]
Looks good, thanks!

**Comment by MikeSpreitzer** [comment]
@kubernetes/sig-api-machinery-bugs

**Comment by spiffxp** [comment]
/cc @liggitt 
I think this is fair game for v1.21 given the approve above but I'm not comfortable making the call, WDYT

Is there a known issue that this fixes?

**Comment by MikeSpreitzer** [comment]
This has not been reported as a bug by users yet, the bug was found by someone working on the code.

**Comment by palnabarun** [comment]
@MikeSpreitzer -- can tests be written which expose the bug that is being solved here? Also, what is the priority for this? Is this critical for v1.22?

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go
Comment (lala123912) [inline pkg/registry/flowcontrol/rest/storage_flowcontrol.go:131]: use structure logging instead. 
```
klog.ErrorS(err, "Ensuring suggested configuration failed")
```

**Comment by liggitt** [comment]
> /cc @liggitt
> I think this is fair game for v1.21 given the approve above but I'm not comfortable making the call, WDYT

I'll defer to @lavalamp / @MikeSpreitzer to determine severity and risk

**Comment by MikeSpreitzer** [comment]
Are there guidelines for judging severity and risk?
This is not something that I expect that happen in normal circumstances.  It can happen if a kube-apiserver is shut down shortly after launch, or is running on a machine that is so heavily loaded that the kube-apiserver does not finish initializing things before it is shut down.
I think @tkashem is best qualified to speak about what can happen when this bug bites.

**Comment by tkashem** [comment]
this bug is unlikely to happen, but given the right condition it can manifest (i have outlined the sequence that triggers this bug). The impact of this bug is that the apiserver can not complete the creation process of the p&f bootstrap configuration and so the cluster operates with an incomplete set of the bootstrap objects.

I am not certain that it has to land in 1.21.0, i think it's okay for it to land in the subsequent patch release. maybe @lavalamp @deads2k can offer some insights.

**Comment by tkashem** [comment]
@MikeSpreitzer this needs a re-lgtm, I had to force push to add structured logging.

**Comment by tkashem** [comment]
when this bug bites, the apiserver will never retry (inside post startup hook) to create the missing `suggested` configuration objects (not without manual intervention). 
The impact can vary depending on which `suggested` configuration objects are missing:
- worst case: no `suggested` configuration object was created successfully, all traffic are assigned to the mandatory `catch-all` configuration. We will see lots of requests being rejected since `catch-all` concurrency share is very limited (`5` only). This will practically render the cluster unusable.
- best case: all `suggested` configuration objects are created except the last one (`global-default` priority level). Some disruption but it's tolerable, I assume.
- average case: half of the `suggested` configuration objects are created (this means we create the suggested flow schema objects but none of the suggested priority level objects are created). As far as impact goes, this is similar to worst case since the dangling flow schema objects will be ignored and all traffic will be allocated to `catch-all` and thus rendering the cluster unusable.

The issue can be resolved manually:
- delete the `mandatory` configuration objects
- restart one of the kube-apiserver instances from the pool.

Given the above, I would recommend putting it land in 1.21.

@MikeSpreitzer thoughts?

**Comment by lavalamp** [comment]
/lgtm
/milestone v1.21

Sounds like an important bug fix to me, and the fix is very low risk.

Seems like we could use some better test but given the severity when this happens, I think we can wait on that.

**Pull request merged**
